modcc/solvers.cpp

```diff
@@ -180,10 +180,25 @@ void SparseSolverVisitor::visit(BlockExpression* e) {
             dvars_.push_back(id->name());
         }
     }
+    scale_factor_.resize(dvars_.size());
 
     BlockRewriterBase::visit(e);
 }
 
+void SparseSolverVisitor::visit(CompartmentExpression *e) {
+    auto loc = e->location();
+
+    for (auto& s: e->is_compartment()->state_vars()) {
+        auto it = std::find(dvars_.begin(), dvars_.end(), s->is_identifier()->spelling());
+        if (it == dvars_.end()) {
+            error({"COMPARTMENT variable is not a state variable", loc});
```

Comment:
I think strictly speaking, we can have a state variable that is not present as a derivative or participant in a kinetic scheme. Maybe an error message such as "COMPARTMENT variable is not used" would be more accurate?

Reply:
Okay, I will update the comment